### ISSUE TITLE
llvm-objcopy: error: unknown argument '--change-section-address'

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#871]
After #779 was resolved for x86_64, I get the following error on 5.6-rc1 with `CONFIG_DEBUG_INFO_BTF`:

```
llvm-objcopy: error: unknown argument '--change-section-address'
llvm-objcopy: error: '.btf.vmlinux.bin': No such file or directory
Failed to generate BTF for vmlinux
Try to disable CONFIG_DEBUG_INFO_BTF
```

This was added in df786c9b947639aedbc7bb44b5dae2a7824af360. Without that commit, the build finishes successfully.

cc @rupprecht

**Comment by nickdesaulniers** [comment]
cc @fomichev @MaskRay

**Comment by nickdesaulniers** [comment]
Filed https://llvm.org/pr45217, too.

**Comment by nickdesaulniers** [comment]
@fomichev has a patch internally I've tested, will be sent upstream shortly.

**Comment by nickdesaulniers** [comment]
submitted: https://lore.kernel.org/bpf/20200316222518.191601-1-sdf@google.com/T/#u

**Comment by MaskRay** [comment]
`defconfig + CONFIG_DEBUG_INFO_BTF=y`

`make CC=clang LD=ld.lld O=/tmp/out/x86_64`

```
+ objcopy --change-section-address .BTF=0 --set-section-flags .BTF=alloc -O binary --only-section=.BTF .tmp_vmlinux.btf .btf.vmlinux.bin
+ objcopy -I binary -O elf64-x86-64 -B x86_64 --rename-section .data=.BTF .btf.vmlinux.bin .btf.vmlinux.bin.o    
objcopy: architecture x86_64 unknown                                       
+ echo 'Failed to generate BTF for vmlinux'
```

The `-B` value is incorrect. It should be `i386:x86_64`. GNU objcopt>=2.34 (https://sourceware.org/bugzilla/show_bug.cgi?id=24968) does not need to specify -B.

**Comment by nathanchance** [comment]
https://lore.kernel.org/netdev/20200318222746.173648-1-maskray@google.com/

should fix `llvm-objcopy` with BTF. I should be able to test in the morning.

**Comment by MaskRay** [comment]
> https://lore.kernel.org/netdev/20200318222746.173648-1-maskray@google.com/
> 
> should fix `llvm-objcopy` with BTF. I should be able to test in the morning.

Thanks! Your test instructions will also be very helpful to me.. I got some trouble because I only started to know how to send a patch at v5.

**Comment by nickdesaulniers** [comment]
accepted: https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=90ceddcb495008ac8ba7a3dce297841efcd7d584

Congrats @MaskRay ! Is that your first Linux kernel patch?

**Comment by MaskRay** [comment]
> accepted: https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=90ceddcb495008ac8ba7a3dce297841efcd7d584
> 
> Congrats @MaskRay ! Is that your first Linux kernel patch?

If bpf-next/master is a -next tree, then yes for the Linux kernel patch merged into a -next tree.

From https://www.kernel.org/doc/html/latest/process/2.Process.html#staging-trees

> When the merge window opens, top-level maintainers will ask Linus to “pull” the patches they have selected for merging from their repositories. If Linus agrees, the stream of patches will flow up into his repository, becoming part of the mainline kernel. The amount of attention that Linus pays to specific patches received in a pull operation varies. It is clear that, sometimes, he looks quite closely. But, as a general rule, Linus trusts the subsystem maintainers to not send bad patches upstream.

Perhaps only after the whole process I can say yes..

**Comment by nathanchance** [comment]
The reason Linux is concerned with this flag has been resolved and merged into mainline: https://git.kernel.org/linus/90ceddcb495008ac8ba7a3dce297841efcd7d584

Closing this here, the upstream LLVM bug can be used to further track implementation of this flag.

**Comment by nickdesaulniers** [comment]
FWIW: the upstream issue was fixed recently via: https://github.com/llvm/llvm-project/commit/2b2f4ae0fbe366ebaa8811bed567a6ff76bbe3b1. Did https://lore.kernel.org/bpf/20200316222518.191601-1-sdf@google.com/ land upstream?


90ceddcb495008ac8ba7a3dce297841efcd7d584 is FWICT what landed upstream from the discussion on https://lore.kernel.org/bpf/20200316222518.191601-1-sdf@google.com/.

**Comment by nathanchance** [comment]
Yup, https://git.kernel.org/linus/90ceddcb495008ac8ba7a3dce297841efcd7d584 in 5.7 avoided this.
